jaxopt/_src/tree_util.py

```diff
@@ -68,3 +73,54 @@ def tree_l2_norm(tree_x, squared=False):
 def tree_zeros_like(tree_x):
   """Creates an all-zero tree with the same structure as tree_x."""
   return tree_map(jnp.zeros_like, tree_x)
+
+
+def tree_collapse(tree, filter_leaves=None):
+  """ Take a pytree of arrays in input and return 1D vector of flattened leaves.
+      Optionnaly accept of pytree `filter_leaves` with same structure as `tree`
+      with a boolean on each leaf. Every `False` leaf are removed from final vector.
+      Reciproqual function of `tree_uncollapse`.   
+  """
+  leaves = tree_leaves(tree)
+  leaves = list(map(lambda params: params.flatten(), leaves))
+  if filter_leaves is not None:
+    filter_leaves = tree_leaves(filter_leaves)
+    leaves = [leaf for leaf, keep in zip(leaves, filter_leaves) if keep]
+  return jnp.concatenate(leaves, axis=0)
+
+
+def tree_uncollapse(tree, vec, filter_leaves=None):
+  """ Take a pytree `tree` of arrays in input, and a 1D vector `vec`.
+      Optionnaly accept a pytree `filter_leaves` with same structure as `tree`
+      having a boolean on each leaf.
+      Transform the 1D vector into a pytree compatible with `tree` structure,
+      with same shapes.
+      `False` leaves are cloned from `tree` 'as is'.
+      Reciproqual function of `tree_collapse`.   
+  """
+  leaves, tree_def = tree_flatten(tree)
+  sizes =  list(map(lambda params: params.size, leaves))
+  shapes = list(map(lambda params: params.shape, leaves))
+  if filter_leaves is not None:
+    boolean_leaves = tree_leaves(filter_leaves)
+    sizes = [size for size, keep in zip(sizes, boolean_leaves) if keep]
+    shapes = [shape for shape, keep in zip(shapes, boolean_leaves) if keep]
+  indices = np.cumsum(np.array(sizes, dtype=np.int32))
+  vecs = jnp.split(vec, indices)
+  vecs = list(map(lambda v,shape: v.reshape(shape), vecs, shapes))
+  if filter_leaves is not None:
+    all_vecs, iter_vecs = [], iter(vecs)
+    for leaf, boolean in zip(leaves, boolean_leaves):
+      all_vecs.append(next(iter_vecs) if boolean else leaf)
+    vecs = all_vecs
+  return tree_unflatten(tree_def, vecs)
+
+
+def tree_select(booleans, tree):
+  """ Accept of pytree of booleans `booleans` with same structure as `tree`.
+      `True` leaves are returned 'as is' while `False` leaves are returned as
```

Comment:
By the way, single quote render as italic while double quotes render like this:

![image](https://user-images.githubusercontent.com/233706/133737233-23e229bb-11a0-474d-85e9-94d0b1ca1ef3.png)